source-s3/main.go

```diff
@@ -106,10 +106,14 @@ func (s *s3Store) List(_ context.Context, query filesource.Query) (filesource.Li
 		input.StartAfter = aws.String(startAt[:len(startAt)-1])
 	}
 
-	return &s3Listing{
+	l := &s3Listing{
 		s3:    s.s3,
 		input: input,
-	}, nil
+	}
+
+	return filesource.ListingFunc(func() (filesource.ObjectInfo, error) {
```

Comment:
It feels a little bit weird to construct an `s3Listing` iterator and then wrap it in a closure. Perhaps add a `ctx` field to `s3Listing` instead?

Reply:
Yeah I was trying perhaps a little too hard to not add a context field to a struct, but in this case it is probably more clear, and [not that big of a deal anyway](https://github.com/golang/go/issues/22602). Updated!

Reply:
Yeah, I agree with that guidance in general but IMO storing the context in a closure is morally equivalent to storing it in the struct proper.

The way to actually keep the spirit of those guidelines would be to refactor the listing iterator interface so it's `.Next(ctx)`, but I didn't recommend that because it looks like other filesource implementations wouldn't be able to use that (for example, see `source-gcs` which takes the context from `List()` and makes that part of its `*storage.ObjectIterator`)